include/dlaf/eigensolver/band_to_tridiag/mc.h

```diff
@@ -705,110 +708,116 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
-               dlaf::internal::transform(policy_hp, init_sweep);
-    copy_tridiag(sweep, std::move(dep));
-
-    const SizeType steps = nrStepsForSweep(sweep, size, b);
-
-    SizeType last_dep_new = 0;
-    for (SizeType step = 0; step < steps;) {
-      // First task might apply less steps to align with the boundaries of the HHR tile v.
-      SizeType nr_steps = steps_per_task - (step == 0 ? (sweep % nb) / b : 0);
-      // Last task only applies the remaining steps
-      nr_steps = std::min(nr_steps, steps - step);
-
-      auto dep_index = std::min(ceilDiv(step + nr_steps, nb / b), last_dep);
+  auto before_sweep = [](SemaphorePtr&& sem) { sem->acquire(); };
 
-      const GlobalElementIndex index_v((sweep / b + step) * b, sweep);
-
-      SizeType set_index = ceilDiv(step, nb / b);
+  auto before_sweep_copy_tridiag = [copy_tridiag_task,
+                                    nb](SemaphorePtr&& sem, SizeType sweep,
+                                        const matrix::Tile<BaseType<T>, Device::CPU>& tile_t) {
+    sem->acquire();
+    copy_tridiag_task(sweep - nb, nb, nb, tile_t);
+  };
 
-      deps[set_index] = dlaf::internal::whenAllLift(nr_steps, w_pipeline(),
-                                                    mat_v.readwrite(dist_v.globalTileIndex(index_v)),
-                                                    dist_v.tileElementIndex(index_v), deps[dep_index]) |
-                        dlaf::internal::transform(policy_hp, cont_sweep) | ex::split();
+  const SizeType sweeps = nrSweeps<T>(size);
+  ex::any_sender<TileVectorPtr> tiles_v;
 
-      last_dep_new = set_index;
-      step += nr_steps;
+  for (SizeType sweep = 0; sweep < sweeps; ++sweep) {
+    auto sem_next = std::make_shared<pika::counting_semaphore<>>(0);
+    ex::unique_any_sender<> dep;
+    if (sweep == 0 || sweep % nb != 0) {
+      dep = ex::just(sem) | dlaf::internal::transform(policy_hp, before_sweep);
     }
-
-    // Limit next sweep to only use valid senders from this sweep.
-    last_dep = last_dep_new;
+    else {
+      const auto tile_index = (sweep - 1) / nb;
+      dep = ex::when_all(ex::just(sem, sweep), mat_trid.readwrite(GlobalTileIndex{tile_index, 0})) |
+            dlaf::internal::transform(policy_hp, before_sweep_copy_tridiag);
+    }
+    const SizeType i = sweep / nb;
+    if (sweep % nb == 0) {
+      tiles_v = ex::when_all_vector(matrix::select(
+                    mat_v, common::iterate_range2d(LocalTileIndex{i, i}, LocalTileSize{n - i, 1}))) |
+                ex::then([](auto&& vector) {
+                  return std::make_shared<typename TileVectorPtr::element_type>(std::move(vector));
+                }) |
+                ex::split();
+    }
+    // TODO needs trasformDetach with policy_hp or the priority come from the dependency?
+    ex::start_detached(ex::when_all(ex::just(sem, sem_next, sweep), tiles_v, std::move(dep)) |
+                       ex::then(run_sweep));
```

Comment:
Is this question about how `then(run_sweep)` will run? If yes, `then` will create an inline continuation on an unspecified execution context. In practice this means it will run on the execution context on the last sender to finish from the `when_all` and even more in practice that means it'll be wherever `tiles_v` completes _or_ on the `policy_hp` context of `dep`. If you want to guarantee that it's on a new high priority thread add a `transform(policy_hp)`. If it's likely to be a small task that should be run as soon as possible keep the `then`.

If the question is something different or you already knew this please ignore this :wink:

Reply:
It partially answer my question.
As `run_sweep` can yield waiting for a semaphore to get ready, I want to be sure that, when the task can continue, it is put in the hp queue.

Reply:
Tasks keep their priority when yielding (except for the special priority `boost`, which gives high priority once and then falls back to normal), so assuming `then` starts on a high priority task it will stay on a high priority task, but given the above there's no guarantee that it'll start on a high priority task.